Integrate Google Cloud Build

shobhitagarwal1612 opened the pull request:
### Things done so far:

- [x] Save gradle cache

- [x] Build apk

- [x] Save apk to cloud bucket (Do we want to enable this or skip this step for now?)

- [x] App version code and name

### Notes
 - Added `secrets.properties` to store maps API key
 - Added a dummy google-services.json needed for remote builds
 - Secrets used by the yaml are saved under GCB trigger settings

gino-m commented on gnd/src/mock/google-services.json:
```diff
@@ -0,0 +1,51 @@
+      "client_info": {
+        "mobilesdk_app_id": "1:123456789012:android:1234567890123456",
+        "android_client_info": {
+          "package_name": "gnddemo1.appspot.com"
```
Let's create a separate project for this? I can set one up.

shobhitagarwal1612 replied:
That would be great!

gino-m commented on gnd/build.gradle:
```diff
@@ -53,8 +53,8 @@ android {
-        versionCode gitVersioner.versionCode
-        versionName gitVersioner.versionName
+        versionCode 1
+        versionName '0.0.1'
```
Can we continue to use auto-versions here?

shobhitagarwal1612 replied:
I temporarily removed this as `gitVersioner` requires `.git` directory to function (which is ignored in .gitignore)

Any ideas to overcome this problem?

shobhitagarwal1612 replied:
One idea is to utilize the .buildenv file added in this PR to track the version count and name.

gino-m replied:
How that work when building directly via command-line with Gradle or Android Studio? 

If it's too difficult to keep, feel free to gitVersioner for now, but let's make sure the version number is defined in at most one place.

shobhitagarwal1612 replied:
Done

gino-m commented on cloudbuild.yaml:
```diff
@@ -0,0 +1,76 @@
+  - name: 'build_cache'
+    path: '/build_cache'
+
+timeout: 1800s
```
Reminder to be sure to format these before finalizing (incl. final LF).

shobhitagarwal1612 commented:
@gino-m 
Thanks for your insightful comments!

> Also, how will the private keys be provided to GCB for running end-to-end tests?

We can provide substitution variables in the GCB console (currently used for storing cloud bucket infos).

shobhitagarwal1612 commented:
@gino-m It is now ready for you to review.

The build is green now and saves generated apks to bucket.
(I'm planning to keep the scope of this PR till this point and add code quality checks/tests in subsequent PRs once this is finalized)

gino-m commented on .gitignore:
```diff
@@ -1,6 +1,6 @@
 .*
 !.gitignore
-gradle/
+!gradle/
```
I previously thought we wouldn't want these in VCS, but just looked into it and found this:
>To make the Wrapper files available to other developers and execution environments you’ll need to check them into version control. 
https://docs.gradle.org/current/userguide/gradle_wrapper.html

So looks like this is the right approach.

gino-m commented on .buildenv:
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env bash
+
+export BUILD_NUM=1
```
Is this file only used by GCB? What happens when building from Android Studio?

shobhitagarwal1612 replied:
> Is this file only used by GCB? 
Yes. `BUILD_NUM` is just a template variable which is currently not being used anywhere. 

> What happens when building from Android Studio or local command-line?
This is loaded into the docker image when the build commences. So, this is not used anywhere else.

This is recommended a pattern at https://github.com/GoogleCloudPlatform/cloud-builders-community for loading environment variables into the build env. The idea is to update this file as needed to replicate exact same environment in GCB.

gino-m commented on README.md:
```diff
@@ -150,3 +150,13 @@ $ keytool -list -v -keystore "$HOME/.android/debug.keystore" \
 
 This includes the API key and URL for your new Firebase project.
+
+### Set up google-cloud-sdk
```
Do developers only need to set this up if they want to build with Google Cloud Build? If so it may be worth changing the heading to "Building with Google Cloud Build (optional)".

shobhitagarwal1612 replied:
Docker image only needs to be updated whenever there's a change in the `Dockerfile`, otherwise not. Thanks for the suggestion! Updated.

gino-m commented on gnd/src/mock/google-services.json:
```diff
@@ -0,0 +1,51 @@
+    }
+  ],
+  "configuration_version": "1"
+}
```
Nit: Please add LFs at end of the file here and throughout.

shobhitagarwal1612 replied:
Done

gino-m commented on .gitignore:
```diff
@@ -1,6 +1,6 @@
+!gradle/
 gnd/debug/
 gnd/release/
 **/google_maps_api.xml
```
How is gnd/src/mock/res/values/google_maps_api.xml included if this rule is present? Should we create an ignore override to include it?

shobhitagarwal1612 replied:
I did `git add -f` to force add `google-services.json` and `google_maps_api.xml`. I'll update the .gitignore too.

shobhitagarwal1612 commented:
@gino-m This is ready for your final review!!!

Things modified:
 - Loads of code cleanup
 - License header
 - Inline and file comments
 - Removed unused docker container
 - Added `secrets.properties` for storing API key to avoid creating useless build variant
 - Removed usage of `google_maps_api.xml` with `secrets.properties`
 - Restored `gitversioner` for app versioning

gino-m commented on README.md:
```diff
@@ -150,3 +148,13 @@ $ keytool -list -v -keystore "$HOME/.android/debug.keystore" \
+ 
+3. gcloud auth login
+  
+4. gcloud config set project [PROJECT_ID]
```
Is there a step missing here? What do you do after setting the project to kick off the build?

shobhitagarwal1612 replied:
This is only needed if we need to update the docker images.
So the next step would be to build and upload the docker image. 
(mentioned in the file docs of cloud-build/cloudbuild.yaml)
`gcloud builds submit --config=cloudbuild.yaml --substitutions=_ANDROID_VERSION=28`

gino-m replied:
So should the title be updated to "Set up..." rather than "Building..."?

shobhitagarwal1612 replied:
That sounds a lot better

shobhitagarwal1612 replied:
Maybe "Setup docker images for Google Cloud Build"?

gino-m replied:
Isn't it the other way around? Updated text a bit; let's revise it more in a separate PR as needed. Thanks!

gino-m commented on cloud-builder/cloudbuild.yaml:
```diff
@@ -0,0 +1,52 @@
+# limitations under the License.
+#
+
+# Configures build of Docker images with Android SDK. To build and upload to Google Cloud:
```
Updated comment; did I get this right?

shobhitagarwal1612 replied:
SGTM!

gino-m commented on cloudbuild.yaml:
```diff
@@ -0,0 +1,111 @@
+# limitations under the License.
+#
+
+# Builds Docker image with Ground APK using Cloud Build. Specifically, executes the following steps:
```
Did I get this right?

shobhitagarwal1612 replied:
I missed to mention this that yaml file is submitted automatically by the github app and the build is  triggered. 
It loads (not builds) the docker image  (from the container registry) and then runs the mentioned steps.

gino-m replied:
Ah, sorry I misunderstood. Updated, hope this is better.

shobhitagarwal1612 replied:
Thank you for your updates!!! It looks a lot better now

gino-m commented on gnd/build.gradle:
```diff
@@ -62,10 +68,13 @@ android {
         release {
             minifyEnabled false
             proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
+            resValue("string", "GOOGLE_MAPS_API_KEY", googleMapsApiKey)
```
Cool trick!

shobhitagarwal1612 replied:
This is a pretty huge change. Everyone who has cloned the project will have to create `secrets.properties` and add their API key.
I couldn't find a better method without a lot of disrupting code changes. Please suggest if you have a better solution.
